{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bf78baec-b586-4778-a4c6-f223eae1480a",
   "metadata": {},
   "source": [
    "## Step 1:\n",
    "Let the program know where to find your tools file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bd11804b-b8ed-436b-9461-61494c015158",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "sys.path.append(os.path.join(os.path.realpath(\"..\"), \"Tools\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8580cc3f-4f39-4e22-99dd-667b50296c3f",
   "metadata": {},
   "source": [
    "## Step 2:\n",
    "Load up all of the necessary packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4308b524-124a-4694-9627-06fdbc3ec563",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading up required packages\n",
    "from plastid import BAMGenomeArray, VariableFivePrimeMapFactory, \\\n",
    "                        GTF2_TranscriptAssembler, GFF3_TranscriptAssembler, \\\n",
    "                        Transcript, ThreePrimeMapFactory, CenterMapFactory\n",
    "from plastid.plotting.plots import *\n",
    "import numpy as np\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "import numpy\n",
    "import math\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import utilities as utils\n",
    "from statsmodels.nonparametric.smoothers_lowess import lowess\n",
    "import warnings\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import csv\n",
    "from scipy.sparse.linalg import lsqr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68c4c2bf-c303-4210-803b-3c3c2cf87bcd",
   "metadata": {},
   "source": [
    "## Step 3:\n",
    "Define the paths to our reference genome and annotation files as well as the path to our data for the Autism Spectrum Disorder (ASD) genes identified by Sfari, the top 200 genes whose regulation is altered by FMRP as determined by RiboDiff, The list of genes and their associated TPM values (used to select for genes with sufficient TPM), and the list of primary transcript isoforms provided by APRIS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "d8a5c172-c52c-40da-a501-0c8476d5bcbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_path = \"/home/keeganfl/Desktop/Work_Fall_2021/Protocol_test/genome/dmel/\"\n",
    "data_path = \"/home/keeganfl/Desktop/measuring_binding_domains/data/\"\n",
    "save_path = \"/home/keeganfl/Desktop/measuring_binding_domains/Extra_tables/TE_tables/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be73b416-ed50-49f3-a074-e0ce1d73a54d",
   "metadata": {},
   "source": [
    "## Step 4: \n",
    "load the transcript files so that we can find the sequence of each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f71f2524-bfbb-4eae-a2ea-d52db062a46c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading up the transcrip[t information.\n",
    "transcripts = list(GTF2_TranscriptAssembler(open(reference_path + \"Drosophila_melanogaster.BDGP6.32.103.gtf\"),return_type=Transcript))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f5ff33b-5d95-42cc-a752-6067b352c493",
   "metadata": {},
   "source": [
    "## Step 5:\n",
    "Create a table that lists all genes and their associated lengths. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a6a6470-2bc9-4627-a5bd-65e37121c7bb",
   "metadata": {},
   "source": [
    "### Right now I am using cds length instead of full lenght, I think that is correct. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "88ac4046-4a82-4e27-bf0c-c459b6d2866f",
   "metadata": {},
   "outputs": [],
   "source": [
    "lengths_cds = []\n",
    "gene_name = []\n",
    "gene_id = []\n",
    "for tran in transcripts:\n",
    "    try:\n",
    "        tran.attr[\"gene_name\"]\n",
    "    except:\n",
    "        continue\n",
    "    try:\n",
    "        lengths_cds.append(tran.cds_end - tran.cds_start)\n",
    "    except:\n",
    "        continue\n",
    "    gene_name.append(tran.attr[\"gene_name\"])\n",
    "    gene_id.append(tran.attr[\"gene_id\"])\n",
    "all_genes = pd.DataFrame(list(zip(gene_name, gene_id, lengths_cds)))\n",
    "all_genes.columns = [\"gene_name\", \"gene_id\", \"lengths_cds\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "feba8953-e26a-40d5-971b-65e264846b6d",
   "metadata": {},
   "source": [
    "## Step 6:\n",
    "Load up the list of APRIS genes. This list will include the transcript IDs of all of the primary isoforms, and will be used to select a single transcript for all genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e7a2dae1-c792-4b1a-91f2-946581ffe92f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load up the Apris genes. \n",
    "prin_trans = pd.read_csv(data_path + \"apris_fly_data.txt\", names = [\"gene_name\", \"???\", \"gene_id\", \"rank\"])\n",
    "prin_trans.drop(columns = [\"???\", \"gene_id\", \"rank\"], inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef09490e-a844-4555-96fa-e0924c2be2c6",
   "metadata": {},
   "source": [
    "## Step 7:\n",
    "Merge the list of APRIS genes with the list of all gene lengths in order to remove the non-primary isoforms. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "1100fef3-9933-4ebc-a233-d9fa502cafbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge the count table with the table of APRIS genes\n",
    "prime_genes = pd.merge(all_genes, prin_trans, how=\"left\", on=\"gene_name\", indicator = True)\n",
    "\n",
    "# Only keep counts with a transcript ID that matched one of the APRIS primary IDs.\n",
    "prime_genes = prime_genes.loc[prime_genes._merge == \"both\"].copy()\n",
    "\n",
    "# Drop any duplicate genes (duplicates may remain if multiple transcripts have identical cds regions).\n",
    "prime_genes.drop_duplicates(subset =\"gene_name\",keep = \"first\", inplace = True) \n",
    "\n",
    "# Remove unnecesary columns. \n",
    "prime_genes.drop(columns = [\"_merge\"], inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f097b75-e034-4603-8cbc-c4ebcf4e408f",
   "metadata": {},
   "source": [
    "## Step 8:\n",
    "Load up the list counts from each example and then merge them with the list of the pimary gene lengths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "75d3df74-832c-4514-93b8-db8d719e821a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load up count data.\n",
    "fly_counts = pd.read_csv(data_path + \"counts/fly_counts.csv\")\n",
    "\n",
    "# Merge with gene length data. \n",
    "len_counts = pd.merge(prime_genes, fly_counts, how = \"inner\", left_on = \"gene_id\", right_on = \"Entry\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea6d5901-dd55-4bf9-a2ed-a2213bfe6b6d",
   "metadata": {},
   "source": [
    "## Step 9: \n",
    "Calculate TPM for each column using the equation \n",
    "<br />\n",
    "$TPM = 10^6 * \\frac{reads\\: mapped\\: to\\: the\\: transcript\\:/\\: transcript\\: length }{Sum(reads\\: mapped\\: to\\: each\\: transcript\\:/\\: each\\: transcript's\\: length)}\n",
    "$\n",
    "<br />\n",
    "Then add them as new columns to the pandas dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "74aa84e2-5a64-4fea-acae-00c55a326f42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate RPK for each gene so that we can use it to calculate TPM. \n",
    "for i in len_counts.columns[4:24]:\n",
    "    rpk = len_counts[i]/len_counts[\"lengths_cds\"]\n",
    "    len_counts[\"%s_rpk\" % (i)] = rpk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "05fc52f6-082a-4ac2-83a5-70fe298c903b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate TPM for each gene. \n",
    "for i in len_counts.columns[4:24]:\n",
    "    per_m_s = sum(len_counts[\"%s_rpk\" % i])/1e6\n",
    "    len_counts[\"%s_tpm\" % (i)] = (len_counts[\"%s_rpk\" % i])/per_m_s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "772ad30b-2925-4305-81f5-b99d00c899b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop all of the RPK columns as we only needed them to get TPM. \n",
    "len_counts.drop(columns = ['RbCtlR1_rpk', 'RbCtlR2_rpk', 'RbCtlR3_rpk', 'RbCtlR4_rpk',\n",
    "       'RbTrtR1_rpk', 'RbTrtR2_rpk', 'RbTrtR3_rpk', 'RbTrtR4_rpk',\n",
    "       'RbTrtR5_rpk', 'RbTrtR6_rpk', 'RbTrtR7_rpk', 'RnaCtlR1_rpk',\n",
    "       'RnaCtlR2_rpk', 'RnaCtlR3_rpk', 'RnaTrtR1_rpk', 'RnaTrtR2_rpk',\n",
    "       'RnaTrtR3_rpk', 'RnaTrtR4_rpk', 'RnaTrtR5_rpk', 'RnaTrtR6_rpk'], inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8688fb33-85fd-4438-aaab-89c58f55b002",
   "metadata": {},
   "source": [
    "## Step 10:\n",
    "Calculate the Average TPM for the mutant and control RNA and RB samples and then calculate average TE. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "883ddf61-50a5-4a56-aca8-edf3d411538e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Avergae tpm for the ribosome profiling controls\n",
    "len_counts['RbCtl_avg_tpm'] = len_counts[['RbCtlR1_tpm', 'RbCtlR2_tpm', 'RbCtlR3_tpm', 'RbCtlR4_tpm']].mean(axis = 1)\n",
    "\n",
    "# Average tpm for ribosome profiling treated\n",
    "len_counts['RbTrt_avg_tpm'] = len_counts[['RbTrtR1_tpm', 'RbTrtR2_tpm', 'RbTrtR3_tpm', 'RbTrtR4_tpm',\n",
    "            'RbTrtR5_tpm', 'RbTrtR6_tpm', 'RbTrtR7_tpm']].mean(axis = 1)\n",
    "\n",
    "# Average tpm for the RNA sequncing control\n",
    "len_counts['RnaCtl_avg_tpm'] = len_counts[['RnaCtlR1_tpm', 'RnaCtlR2_tpm', 'RnaCtlR3_tpm']].mean(axis = 1)\n",
    "\n",
    "# Avergae tpm for the RNA sequencing treated\n",
    "len_counts['RnaTrt_avg_tpm'] = len_counts[['RnaTrtR1_tpm', 'RnaTrtR2_tpm','RnaTrtR3_tpm',\n",
    "            'RnaTrtR4_tpm', 'RnaTrtR5_tpm', 'RnaTrtR6_tpm']].mean(axis= 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "4338e036-efb5-40e2-b834-bab80754f8fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "len_counts['Ctl_TE'] = len_counts['RbCtl_avg_tpm']/len_counts['RnaCtl_avg_tpm']\n",
    "len_counts['Trt_TE'] = len_counts['RbTrt_avg_tpm']/len_counts['RnaTrt_avg_tpm']\n",
    "len_counts['FldChng_TE'] = len_counts['Trt_TE']/len_counts['Ctl_TE']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f6c73c2-309f-42ee-b383-619b59d0a2e7",
   "metadata": {},
   "source": [
    "## Step 11\n",
    "Drop uneccessary columns and save the csv. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "62417b54-ef96-4e69-afe3-44b06a5327fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "len_counts.drop(columns = [\"gene_id\", \"lengths_cds\"], inplace = True)\n",
    "len_counts.to_csv(save_path + \"TE_fly.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "f6e84780-c808-4665-b38f-e7f1c179b214",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Entry</th>\n",
       "      <th>RbCtlR1</th>\n",
       "      <th>RbCtlR2</th>\n",
       "      <th>RbCtlR3</th>\n",
       "      <th>RbCtlR4</th>\n",
       "      <th>RbTrtR1</th>\n",
       "      <th>RbTrtR2</th>\n",
       "      <th>RbTrtR3</th>\n",
       "      <th>RbTrtR4</th>\n",
       "      <th>...</th>\n",
       "      <th>RnaTrtR4_tpm</th>\n",
       "      <th>RnaTrtR5_tpm</th>\n",
       "      <th>RnaTrtR6_tpm</th>\n",
       "      <th>RbCtl_avg_tpm</th>\n",
       "      <th>RbTrt_avg_tpm</th>\n",
       "      <th>RnaCtl_avg_tpm</th>\n",
       "      <th>RnaTrt_avg_tpm</th>\n",
       "      <th>Ctl_TE</th>\n",
       "      <th>Trt_TE</th>\n",
       "      <th>FldChng_TE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>l(2)gl</td>\n",
       "      <td>FBgn0002121</td>\n",
       "      <td>237</td>\n",
       "      <td>441</td>\n",
       "      <td>526</td>\n",
       "      <td>229</td>\n",
       "      <td>551</td>\n",
       "      <td>357</td>\n",
       "      <td>307</td>\n",
       "      <td>189</td>\n",
       "      <td>...</td>\n",
       "      <td>154.223986</td>\n",
       "      <td>136.910824</td>\n",
       "      <td>156.239688</td>\n",
       "      <td>53.425168</td>\n",
       "      <td>56.315812</td>\n",
       "      <td>140.098918</td>\n",
       "      <td>130.886259</td>\n",
       "      <td>0.381339</td>\n",
       "      <td>0.430265</td>\n",
       "      <td>1.128302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ir21a</td>\n",
       "      <td>FBgn0031209</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.021651</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003740</td>\n",
       "      <td>0.013918</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cda5</td>\n",
       "      <td>FBgn0051973</td>\n",
       "      <td>8</td>\n",
       "      <td>12</td>\n",
       "      <td>32</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>14</td>\n",
       "      <td>22</td>\n",
       "      <td>14</td>\n",
       "      <td>...</td>\n",
       "      <td>0.607419</td>\n",
       "      <td>0.580921</td>\n",
       "      <td>0.583157</td>\n",
       "      <td>2.674754</td>\n",
       "      <td>2.791984</td>\n",
       "      <td>0.626519</td>\n",
       "      <td>0.627173</td>\n",
       "      <td>4.269234</td>\n",
       "      <td>4.451699</td>\n",
       "      <td>1.042740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>dbr</td>\n",
       "      <td>FBgn0067779</td>\n",
       "      <td>63</td>\n",
       "      <td>91</td>\n",
       "      <td>89</td>\n",
       "      <td>39</td>\n",
       "      <td>94</td>\n",
       "      <td>44</td>\n",
       "      <td>43</td>\n",
       "      <td>49</td>\n",
       "      <td>...</td>\n",
       "      <td>23.171139</td>\n",
       "      <td>22.104968</td>\n",
       "      <td>22.253418</td>\n",
       "      <td>12.191389</td>\n",
       "      <td>11.234609</td>\n",
       "      <td>28.237695</td>\n",
       "      <td>22.339551</td>\n",
       "      <td>0.431742</td>\n",
       "      <td>0.502902</td>\n",
       "      <td>1.164822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>galectin</td>\n",
       "      <td>FBgn0031213</td>\n",
       "      <td>138</td>\n",
       "      <td>145</td>\n",
       "      <td>176</td>\n",
       "      <td>112</td>\n",
       "      <td>226</td>\n",
       "      <td>116</td>\n",
       "      <td>158</td>\n",
       "      <td>98</td>\n",
       "      <td>...</td>\n",
       "      <td>90.904270</td>\n",
       "      <td>96.589732</td>\n",
       "      <td>87.513229</td>\n",
       "      <td>64.914043</td>\n",
       "      <td>67.697021</td>\n",
       "      <td>99.789370</td>\n",
       "      <td>90.585432</td>\n",
       "      <td>0.650511</td>\n",
       "      <td>0.747328</td>\n",
       "      <td>1.148833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13916</th>\n",
       "      <td>mt:ND4</td>\n",
       "      <td>FBgn0262952</td>\n",
       "      <td>1357</td>\n",
       "      <td>761</td>\n",
       "      <td>1005</td>\n",
       "      <td>686</td>\n",
       "      <td>1467</td>\n",
       "      <td>1245</td>\n",
       "      <td>1749</td>\n",
       "      <td>986</td>\n",
       "      <td>...</td>\n",
       "      <td>296.012285</td>\n",
       "      <td>347.866626</td>\n",
       "      <td>288.702637</td>\n",
       "      <td>401.562408</td>\n",
       "      <td>656.833801</td>\n",
       "      <td>249.855675</td>\n",
       "      <td>295.971960</td>\n",
       "      <td>1.607177</td>\n",
       "      <td>2.219243</td>\n",
       "      <td>1.380833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13917</th>\n",
       "      <td>mt:ND4L</td>\n",
       "      <td>FBgn0013683</td>\n",
       "      <td>565</td>\n",
       "      <td>278</td>\n",
       "      <td>414</td>\n",
       "      <td>379</td>\n",
       "      <td>725</td>\n",
       "      <td>537</td>\n",
       "      <td>743</td>\n",
       "      <td>501</td>\n",
       "      <td>...</td>\n",
       "      <td>8.163787</td>\n",
       "      <td>10.328018</td>\n",
       "      <td>7.902835</td>\n",
       "      <td>818.041533</td>\n",
       "      <td>1393.891656</td>\n",
       "      <td>11.438269</td>\n",
       "      <td>8.284705</td>\n",
       "      <td>71.517947</td>\n",
       "      <td>168.248807</td>\n",
       "      <td>2.352540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13918</th>\n",
       "      <td>mt:ND6</td>\n",
       "      <td>FBgn0013685</td>\n",
       "      <td>530</td>\n",
       "      <td>298</td>\n",
       "      <td>462</td>\n",
       "      <td>402</td>\n",
       "      <td>873</td>\n",
       "      <td>706</td>\n",
       "      <td>961</td>\n",
       "      <td>351</td>\n",
       "      <td>...</td>\n",
       "      <td>3.810586</td>\n",
       "      <td>5.280517</td>\n",
       "      <td>6.205607</td>\n",
       "      <td>465.734996</td>\n",
       "      <td>834.842860</td>\n",
       "      <td>4.770220</td>\n",
       "      <td>5.724264</td>\n",
       "      <td>97.633869</td>\n",
       "      <td>145.842820</td>\n",
       "      <td>1.493773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13919</th>\n",
       "      <td>mt:Cyt-b</td>\n",
       "      <td>FBgn0013678</td>\n",
       "      <td>2907</td>\n",
       "      <td>2990</td>\n",
       "      <td>3141</td>\n",
       "      <td>2346</td>\n",
       "      <td>6194</td>\n",
       "      <td>3874</td>\n",
       "      <td>5443</td>\n",
       "      <td>1715</td>\n",
       "      <td>...</td>\n",
       "      <td>2534.633300</td>\n",
       "      <td>2789.589683</td>\n",
       "      <td>2305.889251</td>\n",
       "      <td>1388.632526</td>\n",
       "      <td>2209.783924</td>\n",
       "      <td>2410.018831</td>\n",
       "      <td>2511.132481</td>\n",
       "      <td>0.576192</td>\n",
       "      <td>0.879995</td>\n",
       "      <td>1.527261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13920</th>\n",
       "      <td>mt:ND1</td>\n",
       "      <td>FBgn0013679</td>\n",
       "      <td>1235</td>\n",
       "      <td>1072</td>\n",
       "      <td>1066</td>\n",
       "      <td>962</td>\n",
       "      <td>1955</td>\n",
       "      <td>1382</td>\n",
       "      <td>1981</td>\n",
       "      <td>1022</td>\n",
       "      <td>...</td>\n",
       "      <td>368.899389</td>\n",
       "      <td>462.914386</td>\n",
       "      <td>349.058122</td>\n",
       "      <td>653.797161</td>\n",
       "      <td>1098.928813</td>\n",
       "      <td>317.707408</td>\n",
       "      <td>366.953435</td>\n",
       "      <td>2.057859</td>\n",
       "      <td>2.994736</td>\n",
       "      <td>1.455268</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13921 rows × 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name        Entry  RbCtlR1  RbCtlR2  RbCtlR3  RbCtlR4  RbTrtR1  \\\n",
       "0        l(2)gl  FBgn0002121      237      441      526      229      551   \n",
       "1         Ir21a  FBgn0031209        0        0        0        0        0   \n",
       "2          Cda5  FBgn0051973        8       12       32        6        6   \n",
       "3           dbr  FBgn0067779       63       91       89       39       94   \n",
       "4      galectin  FBgn0031213      138      145      176      112      226   \n",
       "...         ...          ...      ...      ...      ...      ...      ...   \n",
       "13916    mt:ND4  FBgn0262952     1357      761     1005      686     1467   \n",
       "13917   mt:ND4L  FBgn0013683      565      278      414      379      725   \n",
       "13918    mt:ND6  FBgn0013685      530      298      462      402      873   \n",
       "13919  mt:Cyt-b  FBgn0013678     2907     2990     3141     2346     6194   \n",
       "13920    mt:ND1  FBgn0013679     1235     1072     1066      962     1955   \n",
       "\n",
       "       RbTrtR2  RbTrtR3  RbTrtR4  ...  RnaTrtR4_tpm  RnaTrtR5_tpm  \\\n",
       "0          357      307      189  ...    154.223986    136.910824   \n",
       "1            0        0        0  ...      0.000000      0.000000   \n",
       "2           14       22       14  ...      0.607419      0.580921   \n",
       "3           44       43       49  ...     23.171139     22.104968   \n",
       "4          116      158       98  ...     90.904270     96.589732   \n",
       "...        ...      ...      ...  ...           ...           ...   \n",
       "13916     1245     1749      986  ...    296.012285    347.866626   \n",
       "13917      537      743      501  ...      8.163787     10.328018   \n",
       "13918      706      961      351  ...      3.810586      5.280517   \n",
       "13919     3874     5443     1715  ...   2534.633300   2789.589683   \n",
       "13920     1382     1981     1022  ...    368.899389    462.914386   \n",
       "\n",
       "       RnaTrtR6_tpm  RbCtl_avg_tpm  RbTrt_avg_tpm  RnaCtl_avg_tpm  \\\n",
       "0        156.239688      53.425168      56.315812      140.098918   \n",
       "1          0.021651       0.000000       0.000000        0.003740   \n",
       "2          0.583157       2.674754       2.791984        0.626519   \n",
       "3         22.253418      12.191389      11.234609       28.237695   \n",
       "4         87.513229      64.914043      67.697021       99.789370   \n",
       "...             ...            ...            ...             ...   \n",
       "13916    288.702637     401.562408     656.833801      249.855675   \n",
       "13917      7.902835     818.041533    1393.891656       11.438269   \n",
       "13918      6.205607     465.734996     834.842860        4.770220   \n",
       "13919   2305.889251    1388.632526    2209.783924     2410.018831   \n",
       "13920    349.058122     653.797161    1098.928813      317.707408   \n",
       "\n",
       "       RnaTrt_avg_tpm     Ctl_TE      Trt_TE  FldChng_TE  \n",
       "0          130.886259   0.381339    0.430265    1.128302  \n",
       "1            0.013918   0.000000    0.000000         NaN  \n",
       "2            0.627173   4.269234    4.451699    1.042740  \n",
       "3           22.339551   0.431742    0.502902    1.164822  \n",
       "4           90.585432   0.650511    0.747328    1.148833  \n",
       "...               ...        ...         ...         ...  \n",
       "13916      295.971960   1.607177    2.219243    1.380833  \n",
       "13917        8.284705  71.517947  168.248807    2.352540  \n",
       "13918        5.724264  97.633869  145.842820    1.493773  \n",
       "13919     2511.132481   0.576192    0.879995    1.527261  \n",
       "13920      366.953435   2.057859    2.994736    1.455268  \n",
       "\n",
       "[13921 rows x 49 columns]"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea2310a2-e082-45f0-a785-6923698f0eb4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
