{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bf78baec-b586-4778-a4c6-f223eae1480a",
   "metadata": {},
   "source": [
    "## Step 1:\n",
    "Let the program know where to find your tools file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bd11804b-b8ed-436b-9461-61494c015158",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys, os\n",
    "sys.path.append(os.path.join(os.path.realpath(\"..\"), \"Tools\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8580cc3f-4f39-4e22-99dd-667b50296c3f",
   "metadata": {},
   "source": [
    "## Step 2:\n",
    "Load up all of the necessary packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4308b524-124a-4694-9627-06fdbc3ec563",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading up required packages\n",
    "from plastid import BAMGenomeArray, VariableFivePrimeMapFactory, \\\n",
    "                        GTF2_TranscriptAssembler, GFF3_TranscriptAssembler, \\\n",
    "                        Transcript, ThreePrimeMapFactory, CenterMapFactory\n",
    "from plastid.plotting.plots import *\n",
    "import numpy as np\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "import numpy\n",
    "import math\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import utilities as utils\n",
    "from statsmodels.nonparametric.smoothers_lowess import lowess\n",
    "import warnings\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import csv\n",
    "from scipy.sparse.linalg import lsqr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68c4c2bf-c303-4210-803b-3c3c2cf87bcd",
   "metadata": {},
   "source": [
    "## Step 3:\n",
    "Define the paths to our reference genome and annotation files as well as the path to our data for the Autism Spectrum Disorder (ASD) genes identified by Sfari, the top 200 genes whose regulation is altered by FMRP as determined by RiboDiff, The list of genes and their associated TPM values (used to select for genes with sufficient TPM), and the list of primary transcript isoforms provided by APRIS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d8a5c172-c52c-40da-a501-0c8476d5bcbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_path = \"/home/keeganfl/Desktop/Work_Fall_2021/Protocol_test/genome/mouse/\"\n",
    "data_path = \"/home/keeganfl/Desktop/measuring_binding_domains/data/\"\n",
    "save_path = \"/home/keeganfl/Desktop/measuring_binding_domains/Extra_tables/TE_tables/\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be73b416-ed50-49f3-a074-e0ce1d73a54d",
   "metadata": {},
   "source": [
    "## Step 4: \n",
    "load the transcript files so that we can find the sequence of each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f71f2524-bfbb-4eae-a2ea-d52db062a46c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading up the transcrip[t information.\n",
    "transcripts = list(GTF2_TranscriptAssembler(open(reference_path + \"mm10.refGene.gtf\"),return_type=Transcript))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f5ff33b-5d95-42cc-a752-6067b352c493",
   "metadata": {},
   "source": [
    "## Step 5:\n",
    "Create a table that lists all genes and their associated lengths. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a6a6470-2bc9-4627-a5bd-65e37121c7bb",
   "metadata": {},
   "source": [
    "### Right now I am using cds length instead of full lenght, I think that is correct. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "88ac4046-4a82-4e27-bf0c-c459b6d2866f",
   "metadata": {},
   "outputs": [],
   "source": [
    "lengths_cds = []\n",
    "gene_name = []\n",
    "gene_id = []\n",
    "for tran in transcripts:\n",
    "    try:\n",
    "        tran.attr[\"gene_name\"]\n",
    "    except:\n",
    "        continue\n",
    "    try:\n",
    "        lengths_cds.append(tran.cds_end - tran.cds_start)\n",
    "    except:\n",
    "        continue\n",
    "    gene_name.append(tran.attr[\"gene_name\"])\n",
    "    gene_id.append(tran.attr[\"gene_id\"])\n",
    "all_genes = pd.DataFrame(list(zip(gene_name, gene_id, lengths_cds)))\n",
    "all_genes.columns = [\"gene_name\", \"gene_id\", \"lengths_cds\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "feba8953-e26a-40d5-971b-65e264846b6d",
   "metadata": {},
   "source": [
    "## Step 6:\n",
    "Load up the list of APRIS genes. This list will include the transcript IDs of all of the primary isoforms, and will be used to select a single transcript for all genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "e7a2dae1-c792-4b1a-91f2-946581ffe92f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load up the Apris genes. \n",
    "prin_trans = pd.read_csv(data_path + \"apris_mouse_data.txt\", names = [\"gene_name\", \"???\", \"gene_id\", \"rank\"])\n",
    "prin_trans.drop(columns = [\"???\", \"rank\"], inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef09490e-a844-4555-96fa-e0924c2be2c6",
   "metadata": {},
   "source": [
    "## Step 7:\n",
    "Merge the list of APRIS genes with the list of all gene lengths in order to remove the non-primary isoforms. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "1100fef3-9933-4ebc-a233-d9fa502cafbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge the count table with the table of APRIS genes\n",
    "prime_genes = pd.merge(all_genes, prin_trans, how=\"left\", on=\"gene_name\", indicator = True)\n",
    "\n",
    "# Only keep counts with a transcript ID that matched one of the APRIS primary IDs.\n",
    "prime_genes = prime_genes.loc[prime_genes._merge == \"both\"].copy()\n",
    "\n",
    "# Drop any duplicate genes (duplicates may remain if multiple transcripts have identical cds regions).\n",
    "prime_genes.drop_duplicates(subset =\"gene_name\",keep = \"first\", inplace = True) \n",
    "\n",
    "# Remove unnecesary columns. \n",
    "prime_genes.drop(columns = [\"_merge\"], inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f097b75-e034-4603-8cbc-c4ebcf4e408f",
   "metadata": {},
   "source": [
    "## Step 8:\n",
    "Load up the list counts from each example and then merge them with the list of the pimary gene lengths."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "75d3df74-832c-4514-93b8-db8d719e821a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load up count data.\n",
    "mouse_counts = pd.read_csv(data_path + \"counts/mmus_counts_SIM.csv\")\n",
    "\n",
    "# Merge with gene length data. \n",
    "len_counts = pd.merge(prime_genes, mouse_counts, how = \"inner\", left_on = \"gene_name\", right_on = \"Entry\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea6d5901-dd55-4bf9-a2ed-a2213bfe6b6d",
   "metadata": {},
   "source": [
    "## Step 9: \n",
    "Calculate TPM for each column using the equation \n",
    "<br />\n",
    "$TPM = 10^6 * \\frac{reads\\: mapped\\: to\\: the\\: transcript\\:/\\: transcript\\: length }{Sum(reads\\: mapped\\: to\\: each\\: transcript\\:/\\: each\\: transcript's\\: length)}\n",
    "$\n",
    "<br />\n",
    "Then add them as new columns to the pandas dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "74aa84e2-5a64-4fea-acae-00c55a326f42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate RPK for each gene so that we can use it to calculate TPM. \n",
    "for i in len_counts.columns[5:19]:\n",
    "    rpk = len_counts[i]/len_counts[\"lengths_cds\"]\n",
    "    len_counts[\"%s_rpk\" % (i)] = rpk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "05fc52f6-082a-4ac2-83a5-70fe298c903b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate TPM for each gene. \n",
    "for i in len_counts.columns[5:19]:\n",
    "    per_m_s = sum(len_counts[\"%s_rpk\" % i])/1e6\n",
    "    len_counts[\"%s_tpm\" % (i)] = (len_counts[\"%s_rpk\" % i])/per_m_s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "772ad30b-2925-4305-81f5-b99d00c899b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop all of the RPK columns as we only needed them to get TPM. \n",
    "len_counts.drop(columns = ['RbCtlR1_rpk', 'RbCtlR2_rpk', 'RbCtlR3_rpk',\n",
    "        'RbTrtR1_rpk', 'RbTrtR2_rpk', 'RbTrtR3_rpk',\n",
    "        'RnaCtlR1_rpk','RnaCtlR2_rpk', 'RnaCtlR3_rpk', 'RnaCtlR4_rpk',\n",
    "        'RnaTrtR1_rpk', 'RnaTrtR2_rpk','RnaTrtR3_rpk', 'RnaTrtR4_rpk'], inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8688fb33-85fd-4438-aaab-89c58f55b002",
   "metadata": {},
   "source": [
    "## Step 10:\n",
    "Calculate the Average TPM for the mutant and control RNA and RB samples and then calculate average TE. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "883ddf61-50a5-4a56-aca8-edf3d411538e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Avergae tpm for the ribosome profiling controls\n",
    "len_counts['RbCtl_avg_tpm'] = len_counts[['RbCtlR1_tpm', 'RbCtlR2_tpm', 'RbCtlR3_tpm']].mean(axis = 1)\n",
    "\n",
    "# Average tpm for ribosome profiling treated\n",
    "len_counts['RbTrt_avg_tpm'] = len_counts[['RbTrtR1_tpm', 'RbTrtR2_tpm', 'RbTrtR3_tpm']].mean(axis = 1)\n",
    "\n",
    "# Average tpm for the RNA sequncing control\n",
    "len_counts['RnaCtl_avg_tpm'] = len_counts[['RnaCtlR1_tpm', 'RnaCtlR2_tpm', 'RnaCtlR3_tpm', 'RnaCtlR4_tpm']].mean(axis = 1)\n",
    "\n",
    "# Avergae tpm for the RNA sequencing treated\n",
    "len_counts['RnaTrt_avg_tpm'] = len_counts[['RnaTrtR1_tpm', 'RnaTrtR2_tpm','RnaTrtR3_tpm','RnaTrtR4_tpm']].mean(axis= 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "4338e036-efb5-40e2-b834-bab80754f8fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "len_counts['Ctl_TE'] = len_counts['RbCtl_avg_tpm']/len_counts['RnaCtl_avg_tpm']\n",
    "len_counts['Trt_TE'] = len_counts['RbTrt_avg_tpm']/len_counts['RnaTrt_avg_tpm']\n",
    "len_counts['FldChng_TE'] = len_counts['Trt_TE']/len_counts['Ctl_TE']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dac260d1-ca0b-481b-977e-4801e763c269",
   "metadata": {},
   "source": [
    "## Step 11\n",
    "Drop uneccessary columns and save the csv. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "ddad578d-a044-4785-bde8-66f9533bf56e",
   "metadata": {},
   "outputs": [],
   "source": [
    "len_counts.drop(columns = [\"gene_name\", \"gene_id_x\", \"lengths_cds\"], inplace = True)\n",
    "len_counts.to_csv(save_path + \"TE_mouse.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "f6e84780-c808-4665-b38f-e7f1c179b214",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id_y</th>\n",
       "      <th>Entry</th>\n",
       "      <th>RbCtlR1</th>\n",
       "      <th>RbCtlR2</th>\n",
       "      <th>RbCtlR3</th>\n",
       "      <th>RbTrtR1</th>\n",
       "      <th>RbTrtR2</th>\n",
       "      <th>RbTrtR3</th>\n",
       "      <th>RnaCtlR1</th>\n",
       "      <th>RnaCtlR2</th>\n",
       "      <th>...</th>\n",
       "      <th>RnaTrtR2_tpm</th>\n",
       "      <th>RnaTrtR3_tpm</th>\n",
       "      <th>RnaTrtR4_tpm</th>\n",
       "      <th>RbCtl_avg_tpm</th>\n",
       "      <th>RbTrt_avg_tpm</th>\n",
       "      <th>RnaCtl_avg_tpm</th>\n",
       "      <th>RnaTrt_avg_tpm</th>\n",
       "      <th>Ctl_TE</th>\n",
       "      <th>Trt_TE</th>\n",
       "      <th>FldChng_TE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XM_006495550.3</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>138</td>\n",
       "      <td>202</td>\n",
       "      <td>195</td>\n",
       "      <td>151</td>\n",
       "      <td>183</td>\n",
       "      <td>137</td>\n",
       "      <td>1870</td>\n",
       "      <td>1910</td>\n",
       "      <td>...</td>\n",
       "      <td>55.921464</td>\n",
       "      <td>57.927163</td>\n",
       "      <td>55.148683</td>\n",
       "      <td>23.310738</td>\n",
       "      <td>23.564042</td>\n",
       "      <td>58.021414</td>\n",
       "      <td>56.614057</td>\n",
       "      <td>0.401761</td>\n",
       "      <td>0.416222</td>\n",
       "      <td>1.035995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_011283.2</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>0.051003</td>\n",
       "      <td>0.021386</td>\n",
       "      <td>0.038781</td>\n",
       "      <td>0.292023</td>\n",
       "      <td>0.181681</td>\n",
       "      <td>0.083707</td>\n",
       "      <td>0.056736</td>\n",
       "      <td>3.488617</td>\n",
       "      <td>3.202203</td>\n",
       "      <td>0.917901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XM_011238364.2</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>16</td>\n",
       "      <td>10</td>\n",
       "      <td>15</td>\n",
       "      <td>19</td>\n",
       "      <td>39</td>\n",
       "      <td>67</td>\n",
       "      <td>...</td>\n",
       "      <td>3.768338</td>\n",
       "      <td>1.446195</td>\n",
       "      <td>4.516580</td>\n",
       "      <td>3.464439</td>\n",
       "      <td>4.879884</td>\n",
       "      <td>3.555081</td>\n",
       "      <td>3.411406</td>\n",
       "      <td>0.974503</td>\n",
       "      <td>1.430461</td>\n",
       "      <td>1.467887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001177658.1</td>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>220</td>\n",
       "      <td>257</td>\n",
       "      <td>304</td>\n",
       "      <td>269</td>\n",
       "      <td>252</td>\n",
       "      <td>282</td>\n",
       "      <td>511</td>\n",
       "      <td>495</td>\n",
       "      <td>...</td>\n",
       "      <td>54.767556</td>\n",
       "      <td>49.896480</td>\n",
       "      <td>49.080406</td>\n",
       "      <td>108.263079</td>\n",
       "      <td>127.759845</td>\n",
       "      <td>52.791752</td>\n",
       "      <td>53.382329</td>\n",
       "      <td>2.050757</td>\n",
       "      <td>2.393298</td>\n",
       "      <td>1.167031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_008866.2</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>111</td>\n",
       "      <td>153</td>\n",
       "      <td>150</td>\n",
       "      <td>107</td>\n",
       "      <td>128</td>\n",
       "      <td>136</td>\n",
       "      <td>665</td>\n",
       "      <td>780</td>\n",
       "      <td>...</td>\n",
       "      <td>92.114362</td>\n",
       "      <td>97.057403</td>\n",
       "      <td>88.757785</td>\n",
       "      <td>59.521904</td>\n",
       "      <td>61.150223</td>\n",
       "      <td>75.740772</td>\n",
       "      <td>91.160124</td>\n",
       "      <td>0.785863</td>\n",
       "      <td>0.670800</td>\n",
       "      <td>0.853584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19952</th>\n",
       "      <td>NM_201530.2</td>\n",
       "      <td>Sly</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19953</th>\n",
       "      <td>NM_001017393.3</td>\n",
       "      <td>Gm21943</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19954</th>\n",
       "      <td>XM_017318833.1</td>\n",
       "      <td>LOC108168625</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19955</th>\n",
       "      <td>NM_001160135.2</td>\n",
       "      <td>Gm20806</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19956</th>\n",
       "      <td>NM_001160131.1</td>\n",
       "      <td>Gm20854</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19957 rows Ã— 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id_y         Entry  RbCtlR1  RbCtlR2  RbCtlR3  RbTrtR1  \\\n",
       "0      XM_006495550.3          Xkr4      138      202      195      151   \n",
       "1         NM_011283.2           Rp1        1        6        5        3   \n",
       "2      XM_011238364.2         Sox17        9       11       16       10   \n",
       "3      NM_001177658.1        Mrpl15      220      257      304      269   \n",
       "4         NM_008866.2        Lypla1      111      153      150      107   \n",
       "...               ...           ...      ...      ...      ...      ...   \n",
       "19952     NM_201530.2           Sly        0        0        0        0   \n",
       "19953  NM_001017393.3       Gm21943        0        0        0        0   \n",
       "19954  XM_017318833.1  LOC108168625        0        0        0        0   \n",
       "19955  NM_001160135.2       Gm20806        0        0        0        0   \n",
       "19956  NM_001160131.1       Gm20854        0        0        0        0   \n",
       "\n",
       "       RbTrtR2  RbTrtR3  RnaCtlR1  RnaCtlR2  ...  RnaTrtR2_tpm  RnaTrtR3_tpm  \\\n",
       "0          183      137      1870      1910  ...     55.921464     57.927163   \n",
       "1            1        2         3         4  ...      0.051003      0.021386   \n",
       "2           15       19        39        67  ...      3.768338      1.446195   \n",
       "3          252      282       511       495  ...     54.767556     49.896480   \n",
       "4          128      136       665       780  ...     92.114362     97.057403   \n",
       "...        ...      ...       ...       ...  ...           ...           ...   \n",
       "19952        0        0         0         0  ...      0.000000      0.000000   \n",
       "19953        0        0         0         0  ...      0.000000      0.000000   \n",
       "19954        0        0         0         0  ...      0.000000      0.000000   \n",
       "19955        0        0         0         0  ...      0.000000      0.000000   \n",
       "19956        0        0         0         0  ...      0.000000      0.000000   \n",
       "\n",
       "       RnaTrtR4_tpm  RbCtl_avg_tpm  RbTrt_avg_tpm  RnaCtl_avg_tpm  \\\n",
       "0         55.148683      23.310738      23.564042       58.021414   \n",
       "1          0.038781       0.292023       0.181681        0.083707   \n",
       "2          4.516580       3.464439       4.879884        3.555081   \n",
       "3         49.080406     108.263079     127.759845       52.791752   \n",
       "4         88.757785      59.521904      61.150223       75.740772   \n",
       "...             ...            ...            ...             ...   \n",
       "19952      0.000000       0.000000       0.000000        0.000000   \n",
       "19953      0.000000       0.000000       0.000000        0.000000   \n",
       "19954      0.000000       0.000000       0.000000        0.000000   \n",
       "19955      0.000000       0.000000       0.000000        0.000000   \n",
       "19956      0.000000       0.000000       0.000000        0.000000   \n",
       "\n",
       "       RnaTrt_avg_tpm    Ctl_TE    Trt_TE  FldChng_TE  \n",
       "0           56.614057  0.401761  0.416222    1.035995  \n",
       "1            0.056736  3.488617  3.202203    0.917901  \n",
       "2            3.411406  0.974503  1.430461    1.467887  \n",
       "3           53.382329  2.050757  2.393298    1.167031  \n",
       "4           91.160124  0.785863  0.670800    0.853584  \n",
       "...               ...       ...       ...         ...  \n",
       "19952        0.000000       NaN       NaN         NaN  \n",
       "19953        0.000000       NaN       NaN         NaN  \n",
       "19954        0.000000       NaN       NaN         NaN  \n",
       "19955        0.000000       NaN       NaN         NaN  \n",
       "19956        0.000000       NaN       NaN         NaN  \n",
       "\n",
       "[19957 rows x 37 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "932905f3-f301-48e0-9bc0-580e5657b59d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
